docs/types.rst

```diff
@@ -638,21 +638,27 @@ Another example that uses external function types::
 .. index:: ! type;reference, ! reference type, storage, memory, location, array, struct
 
 Reference Types
-==================
+===============
 
-Complex types, i.e. types which do not always fit into 256 bits have to be handled
-more carefully than the value-types we have already seen. Since copying
-them can be quite expensive, we have to think about whether we want them to be
-stored in **memory** (which is not persisting) or **storage** (where the state
-variables are held).
+Values of reference type can be modified through multiple different names.
+Contrast this with value types where you get an independent copy whenever
+a variable of value type is used. Because of that, reference types have to be handled
+more carefully than value types. Currently, reference types comprise structs,
+arrays and mappings. If you specify a reference type, you always have to explicitly
+provide the data area where the type is stored: ``memory`` (whose lifetime is limited
+to a function call), ``storage`` (the location where the state variables are stored)
+or ``calldata`` (special data location that contains the function arguments,
+only available for external function call parameters).
+
+An assignment that changes the data location will always incur an automatic copy operation,
+while assignments inside the same data location only copy for storage types.
```

Comment:
also only in some cases, isn't it?

Reply:
assignment from storage to storage always copies, unless you assign to a local variable, but yeah, perhaps better to weaken it here and go inte details later.